### PR TITLE
tools/ccache: update to 4.7.2

### PR DESCRIPTION
**Pull request by res0nance**
Update to the latest ccache

Release Notes:
https://ccache.dev/releasenotes.html#_ccache_4_7_2

### CODE REVIEW

**Comment by hauke** [comment]
How did you test this?

**Comment by res0nance** [comment]
> How did you test this?

Enabled ccache and built the repository.

File: tools/ccache/patches/100-honour-copts.patch [hunk 1]
```diff
@@ -1,10 +1,10 @@
...
-                              "OBJCPLUS_INCLUDE_PATH", // clang
-+                             "GCC_HONOUR_COPTS",
-                              nullptr};
-     for (const char** p = envvars; *p; ++p) {
-       const char* v = getenv(*p);
+@@ -1743,6 +1743,7 @@ get_manifest_key(Context& ctx, Hash& hash)
+                            "CPLUS_INCLUDE_PATH",
+                            "OBJC_INCLUDE_PATH",
+                            "OBJCPLUS_INCLUDE_PATH", // clang
++                           "GCC_HONOR_COPTS",
```
Comment (zdz2019) [inline tools/ccache/patches/100-honour-copts.patch:7]: typo? "GCC_HONOUR_COPTS",

Comment (res0nance) [reply tools/ccache/patches/100-honour-copts.patch]: You're right thanks, fixed it

**Comment by trippleflux** [comment]
@res0nance 

Somehow nettle package is always failed to build with ccache 4.7 : 
```
configure: summary of build options:

  Version:           nettle 3.8.1
  Host type:         x86_64-openwrt-linux-gnu
  ABI:               64
  Assembly files:    x86_64/fat x86_64
  Install prefix:    /usr
  Library directory: ${exec_prefix}/lib
  Compiler:          ccache_cc
  Static libraries:  yes
  Shared libraries:  yes
  Public key crypto: yes
  Using mini-gmp:    no
  Documentation:     no

touch /home/username/works/openwrt/build_dir/target-x86_64_glibc_custom/nettle-3.8.1/.configured_a1216157d837c8257d529085967bc205
rm -f /home/username/works/openwrt/build_dir/target-x86_64_glibc_custom/nettle-3.8.1/.built
touch /home/username/works/openwrt/build_dir/target-x86_64_glibc_custom/nettle-3.8.1/.built_check
CFLAGS="-mtune=tigerlake -march=tigerlake -O3 -funsafe-math-optimizations -ftree-vectorize -fno-caller-saves -fno-plt -mtune=tigerlake -march=tigerlake -O3 -funsafe-math-optimizations -ftree-vectorize -fhonour-copts -Wno-error=unused-but-set-variable -Wno-error=unused-result -fmacro-prefix-map=/home/username/works/openwrt/build_dir/target-x86_64_glibc_custom/nettle-3.8.1=nettle-3.8.1 -Wformat -Werror=format-security -fstack-protector -D_FORTIFY_SOURCE=1 -Wl,-z,now -Wl,-z,relro -DPIC -fpic  -I/home/username/works/openwrt/staging_dir/toolchain-x86_64_gcc-12.2.0_glibc/usr/include -I/home/username/works/openwrt/staging_dir/toolchain-x86_64_gcc-12.2.0_glibc/include " CXXFLAGS="-mtune=tigerlake -march=tigerlake -O3 -funsafe-math-optimizations -ftree-vectorize -fno-caller-saves -fno-plt -mtune=tigerlake -march=tigerlake -O3 -funsafe-math-optimizations -ftree-vectorize -fhonour-copts -Wno-error=unused-but-set-variable -Wno-error=unused-result -fmacro-prefix-map=/home/username/works/openwrt/build_dir/target-x86_64_glibc_custom/nettle-3.8.1=nettle-3.8.1 -Wformat -Werror=format-security -fstack-protector -D_FORTIFY_SOURCE=1 -Wl,-z,now -Wl,-z,relro -DPIC -fpic  -I/home/username/works/openwrt/staging_dir/toolchain-x86_64_gcc-12.2.0_glibc/usr/include -I/home/username/works/openwrt/staging_dir/toolchain-x86_64_gcc-12.2.0_glibc/include " LDFLAGS="-L/home/username/works/openwrt/staging_dir/toolchain-x86_64_gcc-12.2.0_glibc/usr/lib -L/home/username/works/openwrt/staging_dir/toolchain-x86_64_gcc-12.2.0_glibc/lib -znow -zrelro " make  -C /home/username/works/openwrt/build_dir/target-x86_64_glibc_custom/nettle-3.8.1/. AR="x86_64-openwrt-linux-gnu-gcc-ar" AS="ccache_cc -c -mtune=tigerlake -march=tigerlake -O3 -funsafe-math-optimizations -ftree-vectorize -fno-caller-saves -fno-plt -mtune=tigerlake -march=tigerlake -O3 -funsafe-math-optimizations -ftree-vectorize -fhonour-copts -Wno-error=unused-but-set-variable -Wno-error=unused-result -fmacro-prefix-map=/home/username/works/openwrt/build_dir/target-x86_64_glibc_custom/nettle-3.8.1=nettle-3.8.1 -Wformat -Werror=format-security -fstack-protector -D_FORTIFY_SOURCE=1 -Wl,-z,now -Wl,-z,relro -DPIC -fpic" LD=x86_64-openwrt-linux-gnu-ld NM="x86_64-openwrt-linux-gnu-gcc-nm" CC="ccache_cc" GCC="ccache_cc" CXX="ccache_cxx" RANLIB="x86_64-openwrt-linux-gnu-gcc-ranlib" STRIP=x86_64-openwrt-linux-gnu-strip OBJCOPY=x86_64-openwrt-linux-gnu-objcopy OBJDUMP=x86_64-openwrt-linux-gnu-objdump SIZE=x86_64-openwrt-linux-gnu-size CROSS="x86_64-openwrt-linux-gnu-" ARCH="x86_64"  DESTDIR="/home/username/works/openwrt/build_dir/target-x86_64_glibc_custom/nettle-3.8.1/ipkg-install" CC="ccache_cc" libnettle.so libhogweed.so
make[3]: Entering directory '/home/username/works/openwrt/build_dir/target-x86_64_glibc_custom/nettle-3.8.1'
/home/username/works/openwrt/staging_dir/host/bin/m4 ./m4-utils.m4 ./asm.m4 config.m4 machine.m4 aes-decrypt-internal.asm >aes-decrypt-internal.s
ccache_cc -I. -I/home/username/works/openwrt/staging_dir/toolchain-x86_64_gcc-12.2.0_glibc/usr/include -I/home/username/works/openwrt/staging_dir/toolchain-x86_64_gcc-12.2.0_glibc/include  -DHAVE_CONFIG_H -mtune=tigerlake -march=tigerlake -O3 -funsafe-math-optimizations -ftree-vectorize -fno-caller-saves -fno-plt -mtune=tigerlake -march=tigerlake -O3 -funsafe-math-optimizations -ftree-vectorize -fhonour-copts -Wno-error=unused-but-set-variable -Wno-error=unused-result -fmacro-prefix-map=/home/username/works/openwrt/build_dir/target-x86_64_glibc_custom/nettle-3.8.1=nettle-3.8.1 -Wformat -Werror=format-security -fstack-protector -D_FORTIFY_SOURCE=1 -Wl,-z,now -Wl,-z,relro -DPIC -fpic  -ggdb3 -Wall -W -Wno-sign-compare   -Wmissing-prototypes -Wmissing-declarations -Wstrict-prototypes   -Wpointer-arith -Wbad-function-cast -Wnested-externs -fpic -MT aes-decrypt-internal.o -MD -MP -MF aes-decrypt-internal.o.d -c aes-decrypt-internal.s
ccache: error: failed to stat aes-decrypt-internal.o.d: No such file or directory
make[3]: *** [Makefile:293: aes-decrypt-internal.o] Error 1
make[3]: Leaving directory '/home/username/works/openwrt/build_dir/target-x86_64_glibc_custom/nettle-3.8.1'
make[2]: *** [Makefile:97: /home/username/works/openwrt/build_dir/target-x86_64_glibc_custom/nettle-3.8.1/.built] Error 2
make[2]: Leaving directory '/home/username/works/openwrt/package/libs/nettle'
time: package/libs/nettle/compile#2.04#1.99#3.75
    ERROR: package/libs/nettle failed to build.
make[1]: *** [package/Makefile:116: package/libs/nettle/compile] Error 1
make[1]: Leaving directory '/home/username/works/openwrt'
make: *** [/home/username/works/openwrt/include/toplevel.mk:231: package/nettle/compile] Error 2
```

Working fine with current ccache 4.6.3

**Comment by res0nance** [comment]
> @res0nance
> 
> Somehow nettle package is always failed to build with ccache 4.7 :
> 
> ```
> configure: summary of build options:
> 
>   Version:           nettle 3.8.1
>   Host type:         x86_64-openwrt-linux-gnu
>   ABI:               64
>   Assembly files:    x86_64/fat x86_64
>   Install prefix:    /usr
>   Library directory: ${exec_prefix}/lib
>   Compiler:          ccache_cc
>   Static libraries:  yes
>   Shared libraries:  yes
>   Public key crypto: yes
>   Using mini-gmp:    no
>   Documentation:     no
> 
> touch /home/username/works/openwrt/build_dir/target-x86_64_glibc_custom/nettle-3.8.1/.configured_a1216157d837c8257d529085967bc205
> rm -f /home/username/works/openwrt/build_dir/target-x86_64_glibc_custom/nettle-3.8.1/.built
> touch /home/username/works/openwrt/build_dir/target-x86_64_glibc_custom/nettle-3.8.1/.built_check
> CFLAGS="-mtune=tigerlake -march=tigerlake -O3 -funsafe-math-optimizations -ftree-vectorize -fno-caller-saves -fno-plt -mtune=tigerlake -march=tigerlake -O3 -funsafe-math-optimizations -ftree-vectorize -fhonour-copts -Wno-error=unused-but-set-variable -Wno-error=unused-result -fmacro-prefix-map=/home/username/works/openwrt/build_dir/target-x86_64_glibc_custom/nettle-3.8.1=nettle-3.8.1 -Wformat -Werror=format-security -fstack-protector -D_FORTIFY_SOURCE=1 -Wl,-z,now -Wl,-z,relro -DPIC -fpic  -I/home/username/works/openwrt/staging_dir/toolchain-x86_64_gcc-12.2.0_glibc/usr/include -I/home/username/works/openwrt/staging_dir/toolchain-x86_64_gcc-12.2.0_glibc/include " CXXFLAGS="-mtune=tigerlake -march=tigerlake -O3 -funsafe-math-optimizations -ftree-vectorize -fno-caller-saves -fno-plt -mtune=tigerlake -march=tigerlake -O3 -funsafe-math-optimizations -ftree-vectorize -fhonour-copts -Wno-error=unused-but-set-variable -Wno-error=unused-result -fmacro-prefix-map=/home/username/works/openwrt/build_dir/target-x86_64_glibc_custom/nettle-3.8.1=nettle-3.8.1 -Wformat -Werror=format-security -fstack-protector -D_FORTIFY_SOURCE=1 -Wl,-z,now -Wl,-z,relro -DPIC -fpic  -I/home/username/works/openwrt/staging_dir/toolchain-x86_64_gcc-12.2.0_glibc/usr/include -I/home/username/works/openwrt/staging_dir/toolchain-x86_64_gcc-12.2.0_glibc/include " LDFLAGS="-L/home/username/works/openwrt/staging_dir/toolchain-x86_64_gcc-12.2.0_glibc/usr/lib -L/home/username/works/openwrt/staging_dir/toolchain-x86_64_gcc-12.2.0_glibc/lib -znow -zrelro " make  -C /home/username/works/openwrt/build_dir/target-x86_64_glibc_custom/nettle-3.8.1/. AR="x86_64-openwrt-linux-gnu-gcc-ar" AS="ccache_cc -c -mtune=tigerlake -march=tigerlake -O3 -funsafe-math-optimizations -ftree-vectorize -fno-caller-saves -fno-plt -mtune=tigerlake -march=tigerlake -O3 -funsafe-math-optimizations -ftree-vectorize -fhonour-copts -Wno-error=unused-but-set-variable -Wno-error=unused-result -fmacro-prefix-map=/home/username/works/openwrt/build_dir/target-x86_64_glibc_custom/nettle-3.8.1=nettle-3.8.1 -Wformat -Werror=format-security -fstack-protector -D_FORTIFY_SOURCE=1 -Wl,-z,now -Wl,-z,relro -DPIC -fpic" LD=x86_64-openwrt-linux-gnu-ld NM="x86_64-openwrt-linux-gnu-gcc-nm" CC="ccache_cc" GCC="ccache_cc" CXX="ccache_cxx" RANLIB="x86_64-openwrt-linux-gnu-gcc-ranlib" STRIP=x86_64-openwrt-linux-gnu-strip OBJCOPY=x86_64-openwrt-linux-gnu-objcopy OBJDUMP=x86_64-openwrt-linux-gnu-objdump SIZE=x86_64-openwrt-linux-gnu-size CROSS="x86_64-openwrt-linux-gnu-" ARCH="x86_64"  DESTDIR="/home/username/works/openwrt/build_dir/target-x86_64_glibc_custom/nettle-3.8.1/ipkg-install" CC="ccache_cc" libnettle.so libhogweed.so
> make[3]: Entering directory '/home/username/works/openwrt/build_dir/target-x86_64_glibc_custom/nettle-3.8.1'
> /home/username/works/openwrt/staging_dir/host/bin/m4 ./m4-utils.m4 ./asm.m4 config.m4 machine.m4 aes-decrypt-internal.asm >aes-decrypt-internal.s
> ccache_cc -I. -I/home/username/works/openwrt/staging_dir/toolchain-x86_64_gcc-12.2.0_glibc/usr/include -I/home/username/works/openwrt/staging_dir/toolchain-x86_64_gcc-12.2.0_glibc/include  -DHAVE_CONFIG_H -mtune=tigerlake -march=tigerlake -O3 -funsafe-math-optimizations -ftree-vectorize -fno-caller-saves -fno-plt -mtune=tigerlake -march=tigerlake -O3 -funsafe-math-optimizations -ftree-vectorize -fhonour-copts -Wno-error=unused-but-set-variable -Wno-error=unused-result -fmacro-prefix-map=/home/username/works/openwrt/build_dir/target-x86_64_glibc_custom/nettle-3.8.1=nettle-3.8.1 -Wformat -Werror=format-security -fstack-protector -D_FORTIFY_SOURCE=1 -Wl,-z,now -Wl,-z,relro -DPIC -fpic  -ggdb3 -Wall -W -Wno-sign-compare   -Wmissing-prototypes -Wmissing-declarations -Wstrict-prototypes   -Wpointer-arith -Wbad-function-cast -Wnested-externs -fpic -MT aes-decrypt-internal.o -MD -MP -MF aes-decrypt-internal.o.d -c aes-decrypt-internal.s
> ccache: error: failed to stat aes-decrypt-internal.o.d: No such file or directory
> make[3]: *** [Makefile:293: aes-decrypt-internal.o] Error 1
> make[3]: Leaving directory '/home/username/works/openwrt/build_dir/target-x86_64_glibc_custom/nettle-3.8.1'
> make[2]: *** [Makefile:97: /home/username/works/openwrt/build_dir/target-x86_64_glibc_custom/nettle-3.8.1/.built] Error 2
> make[2]: Leaving directory '/home/username/works/openwrt/package/libs/nettle'
> time: package/libs/nettle/compile#2.04#1.99#3.75
>     ERROR: package/libs/nettle failed to build.
> make[1]: *** [package/Makefile:116: package/libs/nettle/compile] Error 1
> make[1]: Leaving directory '/home/username/works/openwrt'
> make: *** [/home/username/works/openwrt/include/toplevel.mk:231: package/nettle/compile] Error 2
> ```
> 
> Working fine with current ccache 4.6.3

Hey @trippleflux I managed to reproduce this and found that it was a bug that was fixed in ccache already. Reference: https://github.com/ccache/ccache/issues/1189

I've bumped to the latest ccache and retried it and nettle now compiles without issues.

**Comment by trippleflux** [comment]
Thanks @res0nance , trying 4.7.1 now.

Will edit this message if I am succeeded building pretty large OpenWrt .config .

**Comment by hauke** [comment]
Thank you for your patches, I applied them to master in 06e34a05f9e96d319d426d9bbbfebea9a62cac07, c1c4a5e007db6178a9c7c85e11104d21a7535faf and 7fd764dc5225e853e732cc091533dbaee6dc5419.

**Comment by neheb** [comment]
@hauke so ccache 4.7 mandates std::filesystem, which is only available with g++8, meaning the prereq check needs to be bumped or all these commits reverted.

ping @ynezz @jow-

**Pull request closed without merging**